src/lib/openjp2/segmented_stream.c

```diff
@@ -0,0 +1,335 @@
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "opj_includes.h"
+
+/*--------------------------------------------------------------------------*/
+
+/*  Segmented Buffer Stream */
+
+static void opj_seg_buf_increment(opj_seg_buf_t * seg_buf)
+{
+	opj_buffer_t* cur_seg = NULL;
+	if (seg_buf == NULL ||	seg_buf->cur_seg_id == seg_buf->segments.size-1) {
+		return;
+	}
+
+	cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
+	if (cur_seg->offset == cur_seg->len  && 
+					seg_buf->cur_seg_id < seg_buf->segments.size -1)	{
+		seg_buf->cur_seg_id++;
+	}
+}
+
+static OPJ_SIZE_T opj_seg_buf_read(opj_seg_buf_t * seg_buf,
+									void * p_buffer,
+									OPJ_SIZE_T p_nb_bytes)
+{
+	OPJ_SIZE_T bytes_in_current_segment;
+	OPJ_SIZE_T bytes_to_read;
+	OPJ_SIZE_T total_bytes_read;
+	OPJ_SIZE_T bytes_left_to_read;
+	OPJ_SIZE_T bytes_remaining_in_file;
+
+	if (p_buffer == NULL || p_nb_bytes == 0 || seg_buf == NULL)
+		return 0;
+
+	/*don't try to read more bytes than are available */
+	bytes_remaining_in_file = seg_buf->data_len - opj_seg_buf_get_global_offset(seg_buf);
+	if (p_nb_bytes > bytes_remaining_in_file) {
+		printf("Warning: attempt to read past end of segmented buffer\n");
+		p_nb_bytes = bytes_remaining_in_file;
+	}
+
+	total_bytes_read = 0;
+	bytes_left_to_read = p_nb_bytes;
+	while (bytes_left_to_read > 0 && seg_buf->cur_seg_id < seg_buf->segments.size) {
+		opj_buffer_t* cur_seg = (opj_buffer_t*)opj_vec_get(&seg_buf->segments, seg_buf->cur_seg_id);
+		bytes_in_current_segment = (OPJ_SIZE_T)(cur_seg->len - cur_seg->offset);
+
+		bytes_to_read = (bytes_left_to_read < bytes_in_current_segment) ? 
+											bytes_left_to_read : bytes_in_current_segment;
+
+		if (p_buffer) {
+			memcpy((OPJ_BYTE*)p_buffer + total_bytes_read,cur_seg->buf + cur_seg->offset, bytes_to_read);
+		}
+		opj_seg_buf_incr_cur_seg_offset(seg_buf,bytes_to_read);
+
+		total_bytes_read	+= bytes_to_read;
+		bytes_left_to_read	-= bytes_to_read;
+
+
+	}
+	return total_bytes_read ? total_bytes_read : (OPJ_SIZE_T)-1;
+}
+
+
+static OPJ_OFF_T opj_seg_buf_skip(OPJ_OFF_T p_nb_bytes, opj_seg_buf_t * seg_buf)
+{
+	OPJ_SIZE_T bytes_in_current_segment;
+	OPJ_SIZE_T bytes_remaining;
+
+	if (!seg_buf)
+		return p_nb_bytes;
+
+	if (p_nb_bytes + opj_seg_buf_get_global_offset(seg_buf)> (OPJ_OFF_T)seg_buf->data_len) {
+		printf("Warning: attempt to skip past end of segmented buffer\n");
```

Comment:
There should be no printf in the library code. Please use the existing message mechanism if necessary.